File: frontend/src/services/playerProfileService.ts
// src/services/playerProfileService.ts
// Servizio per la gestione dei profili dei giocatori

import { supabaseClient } from '../lib/supabaseClient';
import { Player, PlayerProfile, PlayerRanking, RANKING_CONFIG } from '../types/player';

export const playerProfileService = {
  // Recupera il profilo di un giocatore
  async getPlayerProfile(userId: string): Promise<PlayerProfile | null> {
    // Prima recupera i dati base del giocatore
    const { data: playerData, error: playerError } = await supabaseClient
      .from('players')
      .select('*')
      .eq('user_id', userId)
      .maybeSingle();
    
    if (playerError) throw new Error(playerError.message);
    if (!playerData) return null;
    
    // Recupera i tornei attivi
    const { data: activeTournaments, error: activeTournamentsError } = await supabaseClient
      .from('tournaments')
      .select('*')
      .eq('creator_user_id', userId)
      .in('status', ['funding_open', 'funding_complete', 'funds_transferred_to_player', 'in_progress', 'awaiting_results'])
      .count();
    
    if (activeTournamentsError) throw new Error(activeTournamentsError.message);
    
    // Recupera i tornei completati
    const { data: completedTournaments, error: completedTournamentsError } = await supabaseClient
      .from('tournaments')
      .select('*')
      .eq('creator_user_id', userId)
      .in('status', ['completed_won', 'completed_lost'])
      .count();
    
    if (completedTournamentsError) throw new Error(completedTournamentsError.message);
    
    // Calcola i guadagni totali (semplificato)
    const { data: earnings, error: earningsError } = await supabaseClient
      .from('tournaments')
      .select('total_winnings_from_tournament')
      .eq('creator_user_id', userId)
      .eq('status', 'completed_won');
    
    if (earningsError) throw new Error(earningsError.message);
    
    const totalEarnings = earnings.reduce((sum, tournament) => sum + (tournament.total_winnings_from_tournament || 0), 0);
    
    return {
      ...playerData as Player,
      total_earnings: totalEarnings,
      active_tournaments: activeTournaments.length,
      completed_tournaments: completedTournaments.length
    };
  },

  // Calcola il ranking di un giocatore in base alle sue statistiche
  calculatePlayerRanking(tournamentsPlayed: number, winRate: number): PlayerRanking {
    if (tournamentsPlayed >= RANKING_CONFIG.PLATINUM.minTournaments && 
        winRate >= RANKING_CONFIG.PLATINUM.minWinRate) {
      return 'PLATINUM';
    } else if (tournamentsPlayed >= RANKING_CONFIG.GOLD.minTournaments && 
               winRate >= RANKING_CONFIG.GOLD.minWinRate) {
      return 'GOLD';
    } else if (tournamentsPlayed >= RANKING_CONFIG.SILVER.minTournaments && 
               winRate >= RANKING_CONFIG.SILVER.minWinRate) {
      return 'SILVER';
    } else {
      return 'BRONZE';
    }
  },

  // Aggiorna il ranking di un giocatore
  async updatePlayerRanking(userId: string): Promise<PlayerRanking> {
    // Recupera i dati del giocatore
    const { data: playerData, error: playerError } = await supabaseClient
      .from('players')
      .select('*')
      .eq('user_id', userId)
      .single();
    
    if (playerError) throw new Error(playerError.message);
    
    // Calcola il nuovo ranking
    const newRanking = this.calculatePlayerRanking(
      playerData.tournaments_played,
      playerData.win_rate
    );
    
    // Aggiorna il ranking nel database
    const { error: updateError } = await supabaseClient
      .from('players')
      .update({ ranking: newRanking, updated_at: new Date().toISOString() })
      .eq('user_id', userId);
    
    if (updateError) throw new Error(updateError.message);
    
    return newRanking;
  },

  // Crea un nuovo profilo giocatore
  async createPlayerProfile(userId: string, name: string, avatarUrl?: string, bio?: string): Promise<Player> {
    const { data, error } = await supabaseClient
      .from('players')
      .insert([
        { 
          user_id: userId,
          name,
          avatar_url: avatarUrl,
          bio,
          ranking: 'BRONZE',
          tournaments_played: 0,
          tournaments_won: 0,
          win_rate: 0
        }
      ])
      .select()
      .single();
    
    if (error) throw new Error(error.message);
    return data as Player;
  },

  // Aggiorna le statistiche di un giocatore dopo un torneo
  async updatePlayerStats(userId: string, won: boolean): Promise<Player> {
    // Recupera i dati attuali del giocatore
    const { data: playerData, error: playerError } = await supabaseClient
      .from('players')
      .select('*')
      .eq('user_id', userId)
      .single();
    
    if (playerError) throw new Error(playerError.message);
    
    // Calcola le nuove statistiche
    const tournamentsPlayed = playerData.tournaments_played + 1;
    const tournamentsWon = won ? playerData.tournaments_won + 1 : playerData.tournaments_won;
    const winRate = tournamentsPlayed > 0 ? tournamentsWon / tournamentsPlayed : 0;
    
    // Aggiorna le statistiche nel database
    const { data, error: updateError } = await supabaseClient
      .from('players')
      .update({ 
        tournaments_played: tournamentsPlayed,
        tournaments_won: tournamentsWon,
        win_rate: winRate,
        updated_at: new Date().toISOString()
      })
      .eq('user_id', userId)
      .select()
      .single();
    
    if (updateError) throw new Error(updateError.message);
    
    // Aggiorna anche il ranking
    await this.updatePlayerRanking(userId);
    
    return data as Player;
  }
};

File: backend/api/services/player_profile_service.py
"""
Player profile service for SolCraft L2.
"""
from typing import Dict, Any, Optional
from uuid import UUID
import logging
from datetime import datetime

logger = logging.getLogger(__name__)

class PlayerProfileService:
    def __init__(self, supabase):
        self.supabase = supabase

    async def get_player_by_user_id(self, user_id: str):
        """Get a player profile by user ID."""
        try:
            response = self.supabase.table("players").select("*").eq("user_id", user_id).single().execute()
            
            if "error" in response:
                return None
            
            return response["data"]
        except Exception as e:
            logger.error(f"Error in get_player_by_user_id: {str(e)}")
            raise

    async def create_player_profile(self, user_id: str, name: str, avatar_url: Optional[str] = None, bio: Optional[str] = None):
        """Create a new player profile."""
        try:
            now = datetime.now().isoformat()
            player_data = {
                "user_id": user_id,
                "name": name,
                "avatar_url": avatar_url,
                "bio": bio,
                "ranking": "BRONZE",
                "tournaments_played": 0,
                "tournaments_won": 0,
                "win_rate": 0,
                "created_at": now,
                "updated_at": now
            }
            
            response = self.supabase.table("players").insert(player_data).execute()
            
            if "error" in response:
                raise Exception(f"Error creating player profile: {response['error']}")
            
            return response["data"][0]
        except Exception as e:
            logger.error(f"Error in create_player_profile: {str(e)}")
            raise

    async def update_player_profile(self, user_id: str, update_data: Dict[str, Any]):
        """Update a player profile."""
        try:
            update_data["updated_at"] = datetime.now().isoformat()
            
            response = self.supabase.table("players").update(update_data).eq("user_id", user_id).execute()
            
            if "error" in response:
                raise Exception(f"Error updating player profile: {response['error']}")
            
            return response["data"][0]
        except Exception as e:
            logger.error(f"Error in update_player_profile: {str(e)}")
            raise

    async def update_player_stats(self, user_id: str, won: bool):
        """Update player statistics after a tournament."""
        try:
            # Recupera il profilo del giocatore
            player = await self.get_player_by_user_id(user_id)
            if not player:
                raise Exception("Player profile not found")
            
            # Aggiorna le statistiche
            tournaments_played = player["tournaments_played"] + 1
            tournaments_won = player["tournaments_won"] + (1 if won else 0)
            win_rate = tournaments_won / tournaments_played if tournaments_played > 0 else 0
            
            # Calcola il nuovo ranking
            new_ranking = self.calculate_ranking(tournaments_played, win_rate)
            
            # Aggiorna il profilo
            update_data = {
                "tournaments_played": tournaments_played,
                "tournaments_won": tournaments_won,
                "win_rate": win_rate,
                "ranking": new_ranking
            }
            
            updated_player = await self.update_player_profile(user_id, update_data)
            
            return updated_player
        except Exception as e:
            logger.error(f"Error in update_player_stats: {str(e)}")
            raise

    def calculate_ranking(self, tournaments_played: int, win_rate: float):
        """Calculate player ranking based on tournaments played and win rate."""
        # Configurazione ranking
        RANKING_CONFIG = {
            "PLATINUM": {
                "minTournaments": 50,
                "minWinRate": 0.65
            },
            "GOLD": {
                "minTournaments": 30,
                "minWinRate": 0.55
            },
            "SILVER": {
                "minTournaments": 15,
                "minWinRate": 0.45
            },
            "BRONZE": {
                "minTournaments": 0,
                "minWinRate": 0
            }
        }
        
        if tournaments_played >= RANKING_CONFIG["PLATINUM"]["minTournaments"] and win_rate >= RANKING_CONFIG["PLATINUM"]["minWinRate"]:
            return "PLATINUM"
        elif tournaments_played >= RANKING_CONFIG["GOLD"]["minTournaments"] and win_rate >= RANKING_CONFIG["GOLD"]["minWinRate"]:
            return "GOLD"
        elif tournaments_played >= RANKING_CONFIG["SILVER"]["minTournaments"] and win_rate >= RANKING_CONFIG["SILVER"]["minWinRate"]:
            return "SILVER"
        else:
            return "BRONZE"

    async def get_player_profile_with_stats(self, user_id: str):
        """Get player profile with additional statistics."""
        try:
            # Recupera il profilo base del giocatore
            player = await self.get_player_by_user_id(user_id)
            if not player:
                return None
            
            # Recupera i tornei attivi
            active_tournaments_response = self.supabase.table("tournaments").select("id").eq("creator_user_id", user_id).in_("status", ["funding_open", "funding_complete", "funds_transferred_to_player", "in_progress", "awaiting_results"]).execute()
            
            # Recupera i tornei completati
            completed_tournaments_response = self.supabase.table("tournaments").select("id").eq("creator_user_id", user_id).in_("status", ["completed_won", "completed_lost"]).execute()
            
            # Calcola i guadagni totali
            earnings_response = self.supabase.table("tournaments").select("total_winnings_from_tournament").eq("creator_user_id", user_id).eq("status", "completed_won").execute()
            
            if "error" not in active_tournaments_response and "error" not in completed_tournaments_response and "error" not in earnings_response:
                active_tournaments = active_tournaments_response["data"]
                completed_tournaments = completed_tournaments_response["data"]
                earnings = earnings_response["data"]
                
                total_earnings = sum(tournament.get("total_winnings_from_tournament", 0) for tournament in earnings)
                
                # Aggiungi le statistiche al profilo
                player["active_tournaments"] = len(active_tournaments)
                player["completed_tournaments"] = len(completed_tournaments)
                player["total_earnings"] = total_earnings
            
            return player
        except Exception as e:
            logger.error(f"Error in get_player_profile_with_stats: {str(e)}")
            raise

    async def list_players(self, limit: int = 50, offset: int = 0):
        """Retrieve a list of player profiles."""
        try:
            response = (
                self.supabase
                .table("players")
                .select("*")
                .range(offset, offset + limit - 1)
                .execute()
            )

            if "error" in response:
                raise Exception(f"Error listing players: {response['error']}")

            return response["data"]
        except Exception as e:
            logger.error(f"Error in list_players: {str(e)}")
            raise
